platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java

```diff
@@ -70,31 +70,47 @@ protected void publish(MavenNormalizedPublication publication, ExternalResourceR
         String version = publication.getVersion();
 
         ModuleArtifactPublisher artifactPublisher = new ModuleArtifactPublisher(repository, localRepo, rootUri, groupId, artifactId, version);
+        SnapshotMetadataResult snapshotMetadataResult = computeSnapshotMetadata(publication, repository, localRepo, version, artifactPublisher, groupId, artifactId);
 
+        if (snapshotMetadataResult != null && !localRepo) {
+            // Use the timestamped version for all published artifacts:
+            // The timestamped version is hidden deep in `Metadata.versioning.snapshotVersions`
+            artifactPublisher.artifactVersion = snapshotMetadataResult.snapshotMetadata.getVersioning().getSnapshotVersions().get(0).getVersion();
```

Comment:
❔ Are there guarantees that `.get(0)` _returns_ the _latest_ snapshot?

Reply:
Refactored the code to isolate this better